### PR TITLE
refactor(compiler): change ngc error handling

### PR DESCRIPTION
**Pull request by bowenni**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Refactoring (no functional changes, no api changes)
```

**What is the current behavior?** (You can also link to an open issue here)
ngc prints stack trace on errors.

**What is the new behavior?**
ngc does not print stack trace on user errors.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

**Other information**:

Do not print stack trace for user errors
Print stack trace for compiler internal errors

### CODE REVIEW

File: modules/@angular/compiler-cli/test/main_spec.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    const mockConsole = {error: {}};
+
+    spyOn(mockConsole, 'error');
+
+    main({p: basePath}, mockConsole.error)
+        .then((exitCode) => {
+          expect(mockConsole.error).toHaveBeenCalled();
+          expect(mockConsole.error).not.toHaveBeenCalledWith('Compilation failed');
+          expect(exitCode).toEqual(0);
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/test/main_spec.ts]: this error code looks wrong, shouldn't it exit 1 in this failure case?

File: modules/@angular/compiler-cli/test/main_spec.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  });
+
+  it('should not print the stack trace if user input file does not exist', (done) => {
+    const mockConsole = {error: {}};
+
+    spyOn(mockConsole, 'error');
+
+    main({p: basePath}, mockConsole.error)
+        .then((exitCode) => {
+          expect(mockConsole.error).toHaveBeenCalled();
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/test/main_spec.ts]: I'm confused, you expect it to have been called? with what? Can we assert the error message contains some string?

File: modules/@angular/compiler-cli/src/main.ts [hunk 3]
```diff
@@ -22,14 +22,24 @@ function codegen(
   return CodeGenerator.create(ngOptions, cliOptions, program, host).codegen();
 }
 
-// CLI entry point
-if (require.main === module) {
-  const args = require('minimist')(process.argv.slice(2));
+export function main(args: any, consoleError: any): Promise<number> {
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/src/main.ts]: a little more idiomatic: give it a type, and make the second arg default to console.error:
`consoleError: (s: string) => void = console.error`

then the caller below just calls with the args argument.

File: modules/@angular/compiler-cli/src/main.ts [hunk 4]
```diff
@@ -22,14 +22,24 @@ function codegen(
...
   const cliOptions = new tsc.NgcCliOptions(args);
-  tsc.main(project, cliOptions, codegen).then(exitCode => process.exit(exitCode)).catch(e => {
-    console.error(e.stack);
-    console.error('Compilation failed');
-    process.exit(1);
+
+  return tsc.main(project, cliOptions, codegen).then(() => 0).catch(e => {
+    if (e instanceof tsc.UserError) {
+      consoleError(e.message);
+      return Promise.resolve(0);
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/src/main.ts]: should be exit 1 in case of any failure (User or not)

**Review by alexeagle — APPROVED** [review]

File: tools/@angular/tsc-wrapped/src/tsc.ts [hunk 5]
```diff
@@ -24,6 +24,15 @@ export interface CompilerInterface {
   emit(program: ts.Program): number;
 }
 
+export class UserError extends Error {
```
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/tsc.ts]: Could we rename this into e.g. `TypecheckError` (or something with `diagnostics` inside)?

File: modules/@angular/compiler-cli/src/main.ts [hunk 6]
```diff
@@ -22,14 +22,25 @@ function codegen(
...
   const project = args.p || args.project || '.';
   const cliOptions = new tsc.NgcCliOptions(args);
-  tsc.main(project, cliOptions, codegen).then(exitCode => process.exit(exitCode)).catch(e => {
-    console.error(e.stack);
-    console.error('Compilation failed');
-    process.exit(1);
+
+  return tsc.main(project, cliOptions, codegen).then(() => 0).catch(e => {
+    if (e instanceof tsc.UserError) {
+      consoleError(e.message);
```
Comment (tbosch) [inline modules/@angular/compiler-cli/src/main.ts]: Add a comment why we don't print `Compilation failed` in this case...

File: tools/@angular/tsc-wrapped/src/tsc.ts (hunk 5 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/tsc.ts]: Ah, I understand now. Never mind, `UserError` is good.

File: modules/@angular/compiler-cli/src/main.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/src/main.ts]: Never mind...

**Pull request merged**